src/Microsoft.OData.Edm/ExtensionMethods/AnnotationsTargetHelper.cs

```diff
@@ -0,0 +1,205 @@
+            string elementName = targetSegments[0];
+
+            IEdmEntityContainer container = model.FindEntityContainer(elementName);
+
+            if (container != null)
+            {
+                results.Add(container);
+
+                if (targetSegmentsCount - 1 > level)
+                {
+                    return HandleIEdmEntityContainer(model, results, container, targetSegments, targetSegmentsCount, level + 1);
+                }
+                else
+                {
+                    return container;
+                }
+            }
+
+            return null;
+        }
+
+        public static IEdmVocabularyAnnotatable HandleIEdmEntityContainer(IEdmModel model, List<IEdmElement> results, IEdmEntityContainer edmEntityContainer, string[] targetSegments, int targetSegmentsCount, int level)
+        {
+            IEdmEntitySet entitySet = edmEntityContainer.FindEntitySetExtended(targetSegments[level]);
+
+            if (entitySet != null)
+            {
+                results.Add(entitySet);
+
+                if (targetSegmentsCount - 1 > level)
+                {
+                    return HandleIEdmContainerElement(model, results, entitySet, targetSegments, targetSegmentsCount, level + 1);
+                }
+                else
+                {
+                    return entitySet;
+                }
+            }
+
+            IEdmSingleton singleton = edmEntityContainer.FindSingletonExtended(targetSegments[level]);
+
+            if (singleton != null)
+            {
+                results.Add(singleton);
+
+                if (targetSegmentsCount - 1 > level)
+                {
+                    return HandleIEdmContainerElement(model, results, singleton, targetSegments, targetSegmentsCount, level + 1);
+                }
+                else
+                {
+                    return singleton;
+                }
+            }
+
+            return null;
+        }
+
+        public static IEdmVocabularyAnnotatable HandleIEdmContainerElement(IEdmModel model, List<IEdmElement> results, IEdmEntityContainerElement edmEntityContainerElement, string[] targetSegments, int targetSegmentsCount, int level)
+        {
+            // .../MyEntitySet/MySchema.MyEntityType/...
+            // .../MyEntitySet/MyComplexProperty/...
+            // .../MySingleton/MyComplexProperty/...
+            IEdmSchemaType type = model.FindType(targetSegments[level]);
+
+            if (type != null)
+            {
+                results.Add(type);
+
+                if (targetSegmentsCount - 1 > level)
+                {
+                    return HandleIEdmSchemaType(model, results, type, targetSegments, targetSegmentsCount, level + 1);
+                }
+                else
+                {
+                    // If we have this path MySchema.MyEntityContainer/MyEntitySet/MySchema.MyEntityType
+                    // Should we return null since it's invalid?
+                    return type;

```

Comment:
>return type; [](http://example.com/codeflow?start=20&length=12)

you should verify the type case is the derived type from previous segment type? It means the type cast is valid or not?